plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/FlushMetadataCacheProcedure.java

```
@@ -0,0 +1,112 @@
+        }
+    }
+
+    @Inject
+    public FlushMetadataCacheProcedure(
+            HiveMetastoreFactory metastoreFactory,
+            Optional<CachingHiveMetastore> cachingHiveMetastore,
+            TransactionLogAccess transactionLogAccess)
+    {
+        this.metastoreFactory = requireNonNull(metastoreFactory, "metastoreFactory is null");
+        this.cachingHiveMetastore = requireNonNull(cachingHiveMetastore, "cachingHiveMetastore is null");
+        this.transactionLogAccess = requireNonNull(transactionLogAccess, "transactionLogAccess is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                PROCEDURE_NAME,
+                ImmutableList.of(
+                        new Procedure.Argument(PARAM_SCHEMA_NAME, VARCHAR, false, null),
+                        new Procedure.Argument(PARAM_TABLE_NAME, VARCHAR, false, null)),
+                FLUSH_HIVE_METASTORE_CACHE.bindTo(this),
+                true);
+    }
+
+    public void flushMetadataCache(ConnectorSession session, String schemaName, String tableName)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doFlushMetadataCache(session, Optional.ofNullable(schemaName), Optional.ofNullable(tableName));
+        }
+    }
+
+    private void doFlushMetadataCache(ConnectorSession session, Optional<String> schemaName, Optional<String> tableName)
+    {
+        if (schemaName.isEmpty() && tableName.isEmpty()) {
+            cachingHiveMetastore.ifPresent(CachingHiveMetastore::flushCache);
+            transactionLogAccess.flushCache();
+        }
+        else if (schemaName.isPresent() && tableName.isPresent()) {
+            HiveMetastore metastore = metastoreFactory.createMetastore(Optional.of(session.getIdentity()));
+            metastore.getTable(schemaName.get(), tableName.get())
+                    .ifPresent(table -> {
```

For a table which doesn't exist, the procedure will still silently succeed. I see the same behaviour for hive's `flush_metadata_cache`. I think this is ok, but want to point it out in case it was not intentional.

---

I followed the existing procedure's behavior, but it might be counterintuitive because we need to get the table location in case of Delta. I changed to throw an exception instead.